matrix-org/matrix-react-sdk
# Better error handling in jump to date

Better error handling in jump to date

 - Friendly error messages with details
 - Add a way to submit debug logs for actual errors (non-networking errors)
 - Don't jump someone back to a room they already navigated away from. Fixes bug mentioned in https://github.com/vector-im/element-web/issues/21263#issuecomment-1056809714

<img width="703" alt="Jump to date error modal" src="https://user-images.githubusercontent.com/558581/226526828-04e6a711-3e36-4945-a908-0adecdb657a4.png">

<img width="699" alt="Jump to date error modal with error details expanded" src="https://user-images.githubusercontent.com/558581/226526830-761fd51c-3157-47a6-84d5-17cf60db2fac.png">

For a non-networking error, we add the option to submit debug logs:
<img width="706" alt="Jump to date error modal with option to submit debug logs for non-networking error" src="https://user-images.githubusercontent.com/558581/226526832-b0daf09c-9910-43c7-b79f-dd77f67eed24.png">





## Testing strategy

 1. Ensure you're using a homeserver with stable support for `/timestamp_to_event` (Synapse 1.73.0+, 2022-12-06)
 1. *Sorry, this next step is hard because [Synapse currently doesn't advertise stable support for Matrix `v.1.6`](https://github.com/matrix-org/synapse/issues/15089) yet (related [Element issue](https://github.com/vector-im/element-web/issues/24362))*
 1. Enable the Labs feature toggle manually by updating the `forcedValue` to `true` in [`src/settings/Settings.tsx#L366-L379`](https://github.com/matrix-org/matrix-react-sdk/blob/ed88e0cdcee5b85e15dec10bf2c80749ef5d83ef/src/settings/Settings.tsx#L366-L379). We can't just enable/disable via `mxSettingsStore` in devtools because "disabled" features can't be enabled/disabled. Here is an example how to edit `Settings.tsx`:
```diff
    "feature_jump_to_date": {
        isFeature: true,
        labsGroup: LabGroup.Messaging,
        displayName: _td("Jump to date (adds /jumptodate and jump to date headers)"),
        supportedLevels: LEVELS_FEATURE,
        default: false,
        controller: new ServerSupportUnstableFeatureController(
            "feature_jump_to_date",
            defaultWatchManager,
            [["org.matrix.msc3030"], ["org.matrix.msc3030.stable"]],
            "v1.6",
            _td("Requires your server to support MSC3030"),
+           true,
        ),
    },
```
 1. Refresh Element so the jump to date headers show up
 1. Open devtools -> Network -> Offline
 1. Open the context menu of one of the jump to date headers and navigate to "Last week"
 1. Notice the error modal that this PR introduces

To see the debug logs option:

 1. To simulate a non-network error, you will have to modify [`src/components/views/messages/DateSeparator.tsx#L127`](https://github.com/matrix-org/matrix-react-sdk/blob/5fefd647c630815901c3f1bc68494b3047d29bbe/src/components/views/messages/DateSeparator.tsx#L127) to `throw new Error('fake error');` and run through the same steps again.


### Dev notes

 - https://kentcdodds.com/blog/common-mistakes-with-react-testing-library
 - `screen.debug()`

### Todo

 - [x] Add test to make sure we don't jump back to room if we already switched away to another room
 - [x] Add test to make sure we don't show an error message for a room if we already switched away to another room


## Checklist

    - Needs exception, see https://github.com/matrix-org/matrix-react-sdk/pull/10405#discussion_r1142916977
-   [ ] ~~Sign-off given on the changes (see [CONTRIBUTING.md](https://github.com/matrix-org/matrix-react-sdk/blob/develop/CONTRIBUTING.md))~~





---
Here's what your changelog entry will look like:

## ✨ Features
 * Better error handling in jump to date ([\#10405](https://github.com/matrix-org/matrix-react-sdk/pull/10405)). Contributed by @MadLittleMods.
